lib/interface_quda.cpp

```diff
@@ -1842,6 +1872,457 @@ void dslashQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+        quda::blas::xpy(pr2, tmp);
+        applySpinTaste(pr2, tmp, gDirs[d2]);
+        myCovDev->MCD (tmp, pr2, d3);
+        myCovDev->MCD (pr1, pr2, d3+4);
+        quda::blas::xpy(pr1, tmp);
+        applySpinTaste(pr1, tmp, gDirs[d3]);
+        myCovDev->MCD (tmp, pr1, d4);
+        myCovDev->MCD (pr2, pr1, d4+4);
+        quda::blas::xpy(pr2, tmp);
+        applySpinTaste(pr2, tmp, gDirs[d4]);
+        quda::blas::xpy(pr2, acc);
+
+        const int m1 = dMnus[i][0];
+        const int m2 = dMnus[i][1];
+        const int m3 = dMnus[i][2];
+        const int m4 = dMnus[i][3];
+
+        // Accumulate result in acc
+        myCovDev->MCD (tmp, out, m1);
+        myCovDev->MCD (pr1, out, m1+4);
+        quda::blas::xpy(pr1, tmp);
+        applySpinTaste(pr1, tmp, gDirs[m1]);
+        myCovDev->MCD (tmp, pr1, m2);
+        myCovDev->MCD (pr2, pr1, m2+4);
+        quda::blas::xpy(pr2, tmp);
+        applySpinTaste(pr2, tmp, gDirs[m2]);
+        myCovDev->MCD (tmp, pr2, m3);
+        myCovDev->MCD (pr1, pr2, m3+4);
+        quda::blas::xpy(pr1, tmp);
+        applySpinTaste(pr1, tmp, gDirs[m3]);
+        myCovDev->MCD (tmp, pr1, m4);
+        myCovDev->MCD (pr2, pr1, m4+4);
+        quda::blas::xpy(pr2, tmp);
+        applySpinTaste(pr2, tmp, gDirs[m4]);
+        quda::blas::mxpy(pr2, acc);
+      }
+
+      applySpinTaste(out, acc, QUDA_SPIN_TASTE_G5);
+      quda::blas::ax(0.0625/24., out);
+      break;
+    }
+  }
+
+  // FIXME: This is not exactly all covDev
+  profileCovDev.TPSTOP(QUDA_PROFILE_COMPUTE);
+
+  profileCovDev.TPSTART(QUDA_PROFILE_D2H);
+  *out_h = out;
+  profileCovDev.TPSTOP(QUDA_PROFILE_D2H);
+
+  if (getVerbosity() >= QUDA_DEBUG_VERBOSE) {
+    double cpu = blas::norm2(*out_h);
+    double gpu = blas::norm2(out);
+    printfQuda("Out CPU %e CUDA %e\n", cpu, gpu);
+  }
+
+  profileCovDev.TPSTART(QUDA_PROFILE_FREE);
+  delete myCovDev; // clean up
+
+  delete out_h;
+  delete in_h;
+  profileCovDev.TPSTOP(QUDA_PROFILE_FREE);
+
+  popVerbosity();
+  profileCovDev.TPSTOP(QUDA_PROFILE_TOTAL);
+}
+
+void covDevQuda(void *h_out, void *h_in, int dir, QudaInvertParam inv_param)
+{
+  profileCovDev.TPSTART(QUDA_PROFILE_TOTAL);
+  profileCovDev.TPSTART(QUDA_PROFILE_INIT);
+
+  const auto &gauge = *gaugePrecise; //(inv_param->dslash_type != QUDA_ASQTAD_DSLASH) ? *gaugePrecise : *gaugeFatPrecise;
+
+  inv_param.solution_type = QUDA_MAT_SOLUTION;
+  inv_param.dirac_order = QUDA_DIRAC_ORDER;
+
+  //if ((!gaugePrecise && inv_param->dslash_type != QUDA_ASQTAD_DSLASH)
+  //    || ((!gaugeFatPrecise || !gaugeLongPrecise) && inv_param->dslash_type == QUDA_ASQTAD_DSLASH))
+  if (!gaugePrecise)
+    errorQuda("Gauge field not allocated");
+
+  pushVerbosity(inv_param.verbosity);
+  if (getVerbosity() >= QUDA_DEBUG_VERBOSE) printQudaInvertParam(&inv_param);
+
+  ColorSpinorParam cpuParam(h_in, inv_param, gauge.X(), false, inv_param.input_location);
+  ColorSpinorField *in_h = ColorSpinorField::Create(cpuParam);
```

Comment:
Move to stack

Reply:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df